src/coreclr/nativeaot/System.Private.CoreLib/src/System/Reflection/Runtime/TypeInfos/RuntimeTypeInfo.cs

```diff
@@ -790,10 +790,37 @@ private TypeClassification Classification
         {
             get
             {
+                // We have a very specialized helper to get the base type.
+                // It is not a general purpose base type, but works for the cases we care about.
+                // This avoids bringing in full type resolution support including constructing
+                // generic types.
+                static Type GetLimitedBaseType(RuntimeTypeInfo thisType)
+                {
+                    // If we have a type handle, just use that
+                    RuntimeTypeHandle typeHandle = thisType.InternalTypeHandleIfAvailable;
+                    if (!typeHandle.IsNull())
+                    {
+                        RuntimeTypeHandle baseTypeHandle;
+                        if (ReflectionCoreExecution.ExecutionEnvironment.TryGetBaseType(typeHandle, out baseTypeHandle))
+                            return Type.GetTypeFromHandle(baseTypeHandle);
+                    }
+
+                    // Metadata fallback. We only care about very limited subset of all possibilities.
```

Comment:
Does this handle `struct` constrains correctly?
```
Console.WriteLine(typeof(G<>).GetGenericArguments()[0].IsValueType); // Prints true

class G<T> where T:struct { }
```

Reply:
Yes because generic parameters implement `TypeRefDefOrSpecForBaseType` as below.

I'll add a dedicated test. We probably have one somewhere, but can't hurt.

https://github.com/dotnet/runtime/blob/12f65cd40c26a0eee613eaccb8c0e53b708b2a8f/src/coreclr/nativeaot/System.Private.CoreLib/src/System/Reflection/Runtime/TypeInfos/RuntimeGenericParameterTypeInfo.cs#L143-L159